File: frontend/src/types/order.ts
export enum OrderStatus {
  PICKED = "picked",
  QA = "qa",
  PRE_DELIVERY = "pre-delivery",
  IN_DELIVERY = "in-delivery",
  SHIPPING = "shipping",
  DELIVERED = "delivered",
  ISSUE = "issue",
}

export enum ShippingWorkflowStatus {
  WORK_AREA = "work_area",
  DOCK = "dock",
  SHIPPED = "shipped",
}

export const ShippingWorkflowStatusDisplayNames: Record<ShippingWorkflowStatus, string> = {
  [ShippingWorkflowStatus.WORK_AREA]: "Work Area",
  [ShippingWorkflowStatus.DOCK]: "At Dock",
  [ShippingWorkflowStatus.SHIPPED]: "Shipped to Carrier",
};

export const OrderStatusDisplayNames: Record<OrderStatus, string> = {
  [OrderStatus.PICKED]: "Picked",
  [OrderStatus.QA]: "QA",
  [OrderStatus.PRE_DELIVERY]: "Pre-Delivery",
  [OrderStatus.IN_DELIVERY]: "In Delivery",
  [OrderStatus.SHIPPING]: "Shipping",
  [OrderStatus.DELIVERED]: "Delivered",
  [OrderStatus.ISSUE]: "Issue",
};

export interface PickStatusItem {
  product_id: string;
  product_name: string;
  ordered: number;
  picked: number;
}

export interface PickStatus {
  is_fully_picked: boolean;
  total_ordered: number;
  total_picked: number;
  missing_items: PickStatusItem[];
}

export interface Order {
  id: string;
  inflow_order_id: string;
  inflow_sales_order_id?: string;
  recipient_name?: string;
  recipient_contact?: string;
  delivery_location?: string;
  po_number?: string;
  status: OrderStatus;
  assigned_deliverer?: string;
  delivery_run_id?: string;
  issue_reason?: string;
  tagged_at?: string;
  tagged_by?: string;
  tag_data?: TagData;
  picklist_generated_at?: string;
  picklist_generated_by?: string;
  picklist_path?: string;
  qa_completed_at?: string;
  qa_completed_by?: string;
  qa_data?: Record<string, any>;
  qa_path?: string;
  qa_method?: string;
  signature_captured_at?: string;
  signed_picklist_path?: string;
  // Shipping workflow fields
  shipping_workflow_status?: ShippingWorkflowStatus;
  shipping_workflow_status_updated_at?: string;
  shipping_workflow_status_updated_by?: string;
  shipped_to_carrier_at?: string;
  shipped_to_carrier_by?: string;
  carrier_name?: string;
  tracking_number?: string;
  inflow_data?: Record<string, any>;
  pick_status?: PickStatus;
  asset_tag_serials?: AssetTagSerial[];
  created_at: string;
  updated_at: string;
}

export interface TagData {
  tag_ids?: string[];
  tag_request_sent_at?: string;
  tag_request_filename?: string;
  tag_request_status?: string;
  [key: string]: any;
}

export interface OrderDetail extends Order {
  inflow_data?: Record<string, any>;
}

export interface AssetTagSerial {
  product_id?: string;
  product_name: string;
  category_id?: string;
  category_name?: string;
  serials: string[];
}

export interface OrderStatusUpdate {
  status: OrderStatus;
  reason?: string;
}

export interface BulkStatusUpdate {
  order_ids: string[];
  status: OrderStatus;
  reason?: string;
}

export interface AuditLog {
  id: string;
  order_id: string;
  changed_by?: string;
  from_status?: string;
  to_status: string;
  reason?: string;
  timestamp: string;
  metadata?: Record<string, any>;
}

export interface TeamsNotification {
  id: string;
  order_id: string;
  teams_message_id?: string;
  sent_at?: string;
  status: "pending" | "sent" | "failed";
  notification_type?: string;
  error_message?: string;
  retry_count: number;
  created_at: string;
}

export interface TeamsConfig {
  webhook_url?: string;
  updated_at: string;
  updated_by?: string;
}

export interface InflowSyncResponse {
  success: boolean;
  orders_synced: number;
  orders_created: number;
  orders_updated: number;
  message: string;
}

File: backend/app/models/order.py
import uuid
from datetime import datetime
import enum

import sqlalchemy as sa
from sqlalchemy import (
    Column,
    String,
    Text,
    DateTime,
    Enum as SQLEnum,
    JSON,
    ForeignKey,
    Index,
    func,
)
from sqlalchemy.orm import relationship, column_property

from app.config import settings
from app.database import Base

_DATABASE_URL = str(settings.database_url).strip().lower()
_IS_MYSQL = _DATABASE_URL.startswith("mysql")


class OrderStatus(str, enum.Enum):
    PICKED = "picked"
    QA = "qa"
    PRE_DELIVERY = "pre-delivery"
    IN_DELIVERY = "in-delivery"
    SHIPPING = "shipping"
    DELIVERED = "delivered"
    ISSUE = "issue"

    @property
    def display_name(self) -> str:
        return {
            "picked": "Picked",
            "qa": "QA",
            "pre-delivery": "Pre-Delivery",
            "in-delivery": "In Delivery",
            "shipping": "Shipping",
            "delivered": "Delivered",
            "issue": "Issue",
        }.get(self.value, self.value)


class ShippingWorkflowStatus(str, enum.Enum):
    WORK_AREA = "work_area"
    DOCK = "dock"
    SHIPPED = "shipped"

    @property
    def display_name(self) -> str:
        return {
            "work_area": "Work Area",
            "dock": "At Dock",
            "shipped": "Shipped to Carrier",
        }.get(self.value, self.value)


class Order(Base):
    __tablename__ = "orders"

    __table_args__ = (
        Index("ix_orders_created_at", "created_at"),
        Index("ix_orders_updated_at", "updated_at"),
        Index("ix_orders_signature_captured_at", "signature_captured_at"),
        Index("ix_orders_status_updated_at", "status", "updated_at"),
        Index(
            "ix_orders_status_tagged_at_updated_at", "status", "tagged_at", "updated_at"
        ),
        Index(
            "ix_orders_delivery_run_id_delivery_sequence",
            "delivery_run_id",
            "delivery_sequence",
        ),
    )

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    inflow_order_id = Column(
        String(255), unique=True, nullable=False, index=True
    )  # e.g., "TH3270"
    if _IS_MYSQL:
        inflow_order_id_lower = Column(
            String(255),
            sa.Computed("lower(inflow_order_id)", persisted=True),
            index=True,
        )
    else:
        inflow_order_id_lower = column_property(func.lower(inflow_order_id))
    inflow_sales_order_id = Column(String(255), nullable=True)  # UUID from Inflow
    recipient_name = Column(String(255), nullable=True)
    recipient_contact = Column(String(255), nullable=True)  # email
    delivery_location = Column(
        String(500), nullable=True
    )  # building/room or shipping address
    po_number = Column(String(255), nullable=True)
    status = Column(
        String(50), nullable=False, default=OrderStatus.PICKED.value, index=True
    )
    assigned_deliverer = Column(String(255), nullable=True)
    issue_reason = Column(Text, nullable=True)
    tagged_at = Column(DateTime, nullable=True)
    tagged_by = Column(String(255), nullable=True)
    tag_data = Column(JSON, nullable=True)
    picklist_generated_at = Column(DateTime, nullable=True)
    picklist_generated_by = Column(String(255), nullable=True)
    picklist_path = Column(String(500), nullable=True)
    delivery_run_id = Column(
        String(36), ForeignKey("delivery_runs.id"), nullable=True, index=True
    )
    delivery_sequence = Column(sa.Integer, nullable=True)
    qa_completed_at = Column(DateTime, nullable=True)
    qa_completed_by = Column(String(255), nullable=True)
    qa_data = Column(JSON, nullable=True)
    qa_path = Column(String(500), nullable=True)
    qa_method = Column(String(50), nullable=True)  # "Delivery" or "Shipping"
    signature_captured_at = Column(DateTime, nullable=True)
    signed_picklist_path = Column(String(500), nullable=True)
    # Order Details PDF fields
    order_details_path = Column(String(500), nullable=True)
    order_details_generated_at = Column(DateTime, nullable=True)
    # Shipping workflow fields
    shipping_workflow_status = Column(
        String(50), nullable=True, default=ShippingWorkflowStatus.WORK_AREA.value
    )
    shipping_workflow_status_updated_at = Column(DateTime, nullable=True)
    shipping_workflow_status_updated_by = Column(String(255), nullable=True)
    shipped_to_carrier_at = Column(DateTime, nullable=True)
    shipped_to_carrier_by = Column(String(255), nullable=True)
    carrier_name = Column(String(100), nullable=True)  # "FedEx", "UPS", etc.
    tracking_number = Column(String(255), nullable=True)
    inflow_data = Column(JSON, nullable=True)  # Full Inflow payload
    # Remainder order tracking
    parent_order_id = Column(
        String(36),
        ForeignKey("orders.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )  # If this is a remainder, points to parent
    has_remainder = Column(
        String(1), nullable=True, default=None
    )  # 'Y' if this order has a remainder
    remainder_order_id = Column(
        String(36),
        ForeignKey("orders.id", ondelete="SET NULL"),
        nullable=True,
    )  # Points to the remainder order
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(
        DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow
    )

    # Relationships
    audit_logs = relationship(
        "AuditLog", back_populates="order", cascade="all, delete-orphan"
    )
    status_history = relationship(
        "OrderStatusHistory", back_populates="order", cascade="all, delete-orphan"
    )
    print_jobs = relationship(
        "PrintJob",
        back_populates="order",
        cascade="all, delete-orphan",
        order_by="desc(PrintJob.created_at)",
    )

    delivery_run = relationship("DeliveryRun", back_populates="orders")

    @property
    def latest_picklist_print_job(self):
        for job in self.print_jobs or []:
            if job.document_type == "picklist":
                return job
        return None
